src/circuit_proof/r1cs.rs

```diff
@@ -0,0 +1,667 @@
+use rand::{CryptoRng, Rng};
+
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use generators::PedersenGenerators;
+use std::iter::FromIterator;
+
+use circuit_proof::{Circuit, ProverInput, VerifierInput};
+
+// This is a stripped-down version of the Bellman r1cs representation, for the purposes of
+// learning / understanding. The goal is to write this as a BulletproofsConstraintSystem that
+// implements the Bellman ConstraintSystem trait, so we can use that code/logic.
+// (That would require the bellman code to be decoupled from the underlying pairings.)
+
+/// Represents a variable in our constraint system, where the value represents the index.
+#[derive(Clone, Debug)]
+pub struct Variable(usize);
+
+impl Variable {
+    pub fn get_index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Represents a linear combination of some variables multiplied with their scalar coefficients,
+/// plus a scalar. E.g. LC = variable[0]*scalar[0] + variable[1]*scalar[1] + scalar
+pub struct LinearCombination {
+    variables: Vec<(Variable, Scalar)>,
+    constant: Scalar,
+}
+
+impl LinearCombination {
+    // TODO: make constructor with iterators
+    // see FromIterator trait - [(a1, v1), (a2, v2)].iter().collect() (pass in the iterator, collect to get LC)
+    pub fn construct(variables: Vec<(Variable, Scalar)>, constant: Scalar) -> Self {
```

Comment:
I think this should probably be called `new` to be consistent with other Rust constructors